runtime-modules/referendum/src/lib.rs

```diff
@@ -46,15 +47,24 @@ impl<BlockNumber, VotePower> Default for ReferendumStage<BlockNumber, VotePower>
 /// Representation for voting stage state.
 #[derive(Encode, Decode, PartialEq, Eq, Debug, Default)]
 pub struct ReferendumStageVoting<BlockNumber> {
-    started: BlockNumber,     // block in which referendum started
-    extra_options_count: u64, // number of options that exceeding the number of winners
+    started: BlockNumber,      // block in which referendum started
+    winning_target_count: u64, // target number of winners
 }
 
 /// Representation for revealing stage state.
 #[derive(Encode, Decode, PartialEq, Eq, Debug, Default)]
 pub struct ReferendumStageRevealing<BlockNumber, VotePower> {
-    started: BlockNumber,                 // block in which referendum started
-    intermediate_results: Vec<VotePower>, // votes that options have recieved at a given time
+    started: BlockNumber,      // block in which referendum started
+    winning_target_count: u64, // target number of winners
+    intermediate_winners: Vec<OptionResult<VotePower>>, // intermediate winning options
+    //intermediate_results: HashMap<u64, VotePower>, // votes that options have recieved at a given time
+    intermediate_results: BTreeMap<u64, VotePower>, // votes that options have recieved at a given time
```

Comment:
I am assuming this holds the amount of staking backing a candidate?

If so, then we cannot represent it like this safely for two reasons
- the number of alternatives, i.e. mappings in this map, is unconstrained. That is in fact the whole point of the change we are making:  not to cap. the number of alternatives. Storing it as a map inline like this will be an attack vector.
- the number of votes per alternative is unconstrained.

Reply:
Please let me know what you think could be an alternative approach prior to staring implementing improvement.

Reply:
To clarify: 
- `intermediate_winners` is a vector of length less than or equal `winning_target_count` (`0 <= intermediate_winners.len() < winning_target_count`) representing ordered list of currently winning options. Since `winning_target_count` is parameter received at the start of referendum, it is up to calling module (e.g. council module) to constrain this value.
- `intermediate_results` is a map `option_id -> sum_of_votes_for_option`; possible `option_id`s can be restricted via `is_valid_option_id(...)`. The values are needed for the time when a user is voting for an option currently not listed among winners so that new total votes for the option can be compared to the current winners. This map never needs to be iterated over in referendum; thus, it shouldn't be an attack vector (or am I wrong here?). If it needs to be iterated in the council module (or others), it is up to that module to constraint the number of options.

Reply:
- `intermediate_winners`: yes, I have no problem with this, I was only commenting on `intermediate_results`.
- `intermediate_results`: yes, despite not being explicitly iterated by you, a BTreeMap is not a normal storage map. While it is indeed a map, it is encoded as a single value in the underlying database. This means _any_ attempt to read, update, count, or do anything with a btreemap involves reading all mappings from the database. This is not the case with a storage map, which is a genuine map in the underlying database model

Reply:
Fixed in  #212fb67.